rfd/0031-dynamic-app-db-registration.md

```diff
@@ -0,0 +1,355 @@
+---
+authors: Roman Tkachenko (roman@goteleport.com)
+state: draft
+---
+
+# RFD 31 - Dynamic Registration for Apps and Databases
+
+## What
+
+Provide ability for users to register/deregister web apps (for App Access)
+and databases (for Database Access) without having to modify static yaml
+configuration or spin up/bring down app/database agents.
+
+## Why
+
+Currently, users have two options for adding/removing web apps and databases
+in a Teleport cluster:
+
+- Update yaml configuration of a particular `app_service` or `db_service`
+  to add/remove an app or a database entry and restart the service.
+- Bring up a new instance of an `app_service` or a `db_service` agent to
+  register a new app/database, or stop an existing one to deregister.
+
+This is inconvenient for multiple reasons:
+
+- Updating static yaml configuration and restarting the agent is manual and
+  affects all other services served by this agent (incl. other apps/databases).
+- Bringing up or stopping an app/database service agent can be challenging
+  from operational perspective.
+- It is unfriendly to automation and from the integration standpoint i.e. no
+  easy way for a CI tool or an external plugin to connect a new app/database.
+- Similar to above, it makes it impossible for tools like our Terraform
+  provider to manage apps/databases as resources.
+
+## Scope
+
+Given the shortcomings of the current approach, this RFD proposes the way to:
+
+- Allow Teleport users with appropriate permissions to add/remove web apps and
+  databases using CLI and API.
+
+The following is **out of scope** of this RFD and/or initial implementation:
+
+- Providing ability to add/remove web apps and databases via Teleport web UI.
```

Comment:
this can be out of scope for initial implementation, but why not include it in the RFD?